runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -156,6 +154,172 @@ static TR::Node *lowerCASValues(
+      {
+      if (trace)
+         traceMsg(comp, "Storing constant 1 in register %s\n", comp->getDebug()->getGlobalRegisterName(anchoredNode->getGlobalRegisterNumber()));
+      storeNode = TR::Node::create(TR::iRegStore, 1, const1Node);
+      storeNode->setGlobalRegisterNumber(anchoredNode->getGlobalRegisterNumber());
+      }
+   else if (anchoredNode->getOpCodeValue() == TR::iload)
+      {
+      if (trace)
+         traceMsg(comp, "Storing constant 1 to symref %d (%s)\n", anchoredNode->getSymbolReference()->getReferenceNumber(), anchoredNode->getSymbolReference()->getName(comp->getDebug()));
+      storeNode = TR::Node::create(TR::istore, 1, const1Node);
+      storeNode->setSymbolReference(anchoredNode->getSymbolReference());
+      }
+   else
+      TR_ASSERT_FATAL(false, "Anchord call has been turned into unexpected opcode %s\n", anchoredNode->getOpCode().getName());
+   prevBlock->append(TR::TreeTop::create(comp, storeNode));
+
+   // instert acmpeq for fastpath, taking care to set the proper register dependencies
+   auto* ifacmpeqNode = TR::Node::createif(TR::ifacmpeq, anchoredCallArg1TT->getNode()->getFirstChild(), anchoredCallArg2TT->getNode()->getFirstChild(), targetBlock->getEntry());
+   if (anchoredNode->getOpCodeValue() == TR::iRegLoad)
+      {
+      auto* depNode = TR::Node::create(TR::PassThrough, 1, storeNode->getChild(0));
+      depNode->setGlobalRegisterNumber(storeNode->getGlobalRegisterNumber());
+
+      TR::Node* glRegDeps = TR::Node::create(TR::GlRegDeps);
+      glRegDeps->addChildren(&depNode, 1);
+      ifacmpeqNode->addChildren(&glRegDeps, 1);
+
+      if (callBlock->getExit()->getNode()->getNumChildren() > 0)
+         {
+         TR::Node* expectedDeps = callBlock->getExit()->getNode()->getFirstChild();
+         for (int i = 0; i < expectedDeps->getNumChildren(); ++i)
+            {
+            TR::Node* temp = expectedDeps->getChild(i);
+            if (temp->getGlobalRegisterNumber() == depNode->getGlobalRegisterNumber())
+               continue;
+            glRegDeps->addChildren(&temp, 1);
+            }
+         }
+      }
+   prevBlock->append(TR::TreeTop::create(comp, ifacmpeqNode));
+   }
+
+void
+J9::CodeGenerator::lowerNonhelperCallIfNeeded(TR::Node *node, TR::TreeTop *tt)
+   {
+   TR::Compilation* comp = self()->comp();
+
+   if (TR::Compiler->om.areValueTypesEnabled() &&
+       comp->getSymRefTab()->isNonHelper(
+       node->getSymbolReference(),
+       TR::SymbolReferenceTable::objectEqualityComparisonSymbol))
+      {
+      // turn the non-helper call into a VM helper call
+      node->setSymbolReference(comp->getSymRefTab()->findOrCreateAcmpHelperSymbolRef());
+      static const bool disableAcmpFastPath =  NULL != feGetEnv("TR_DisableAcmpFastpath");
+      static const bool traceFastPath = NULL != feGetEnv("TR_TraceAcmpFastPath");
+      if (!disableAcmpFastPath)
+         {
+         self()->fastpathAcmpHelper(node, tt, traceFastPath);
```

Comment:
should this trace turn on under traceCG for convenience?

Reply:
Fixed in force push.